### PR TITLE
Display current difficulty statistics with mods applied

### PR DESCRIPTION
**Pull request by LeNitrous**
Allows `BeatmapDetailArea` to display the most current difficulty settings from mods that alter difficulty. This also allows `IApplicableToBeatmapDifficulty` to be called when a new beatmap is selected in `SongSelect`.

No Mod
image
Easy Mod
image
Hard Rock Mod
image
Colors are based off of stable.

### CODE REVIEW

**Comment by peppy** [comment]
May be worth looking at https://github.com/ppy/osu/pull/3272 too.

File: osu.Game/Screens/Select/Details/AdvancedStats.cs [hunk 1]
```diff
@@ -65,12 +60,60 @@ public AdvancedStats()
...
+            mods.ValueChanged += _ => updateStatistics();
+        }
+
+        private void updateStatistics()
+        {
+            BeatmapInfo processed = Beatmap?.Clone();
+
+            if (processed != null && mods.Value.Any(m => m is IApplicableToDifficulty))
+            {
+                processed.BaseDifficulty = processed.BaseDifficulty.Clone();
```
Comment (peppy) [inline osu.Game/Screens/Select/Details/AdvancedStats.cs]: is it necessary to clone a second time?

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
It would be nice if you could see the actual change in difficulty vs the original. Maybe show the adjusted difficulty in the bottom-half of each line (using the colours you have)?

image

Or use the full height but keep the white and put the colour behind (for harder) or in front (for easier) so you can see the difference

image

Should be considered a temporary solution either way (fine to implement) until we can get a design for this.

File: osu.Game/Screens/Select/Details/AdvancedStats.cs (hunk 1 above)
Comment (LeNitrous) [reply osu.Game/Screens/Select/Details/AdvancedStats.cs]: For some reason we need to get a shallow clone of `BeatmapDifficulty`. Certain issues such as the bars not getting colored and the difficulty not showing correct values when re-entering song select will show up.

It is also done in `WorkingBeatmap`:https://github.com/ppy/osu/blob/0627c814d66c34925057fa38a5dc50bd33144d5e/osu.Game/Beatmaps/WorkingBeatmap.cs#L124-L125

**Comment by LeNitrous** [comment]
I've considered using the last suggestion where there is one bar in front with transparency since this is just a temporary solution. I've also updated the value text to change color as well as a quick addition.

The screenshots have been updated to reflect this.

**Comment by peppy** [comment]
I made a couple of fixes to this which I recommend you check out:

- Removed the beatmap cloning – only the difficulty needed to be cloned (for now).
- Used tuple for settings. The way you had it, it would have displayed incorrectly if the Base/Mod value were set back-to-front. By combining into a tuple it not only cleans up the code but ensure a stable state when arriving in the display context.

**Review by peppy — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**